File: cpp/0031.0_Next_Permutation.cpp
/*
执行用时：8 ms, 在所有 C++ 提交中击败了21.93% 的用户
内存消耗：11.7 MB, 在所有 C++ 提交中击败了72.90% 的用户
通过测试用例：265 / 265
*/


class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        next_permutation(nums.begin(), nums.end());
    }
};

File: problems/0031.0_Next_Permutation.py
'''
two pointers
T: O(2N)
S: O(1)

Runtime: 66 ms, faster than 37.56% of Python3 online submissions for Next Permutation.
Memory Usage: 13.9 MB, less than 79.00% of Python3 online submissions for Next Permutation.
'''
'''
4 2 1 6 8 [4] 7 6 5 4 3 1
            7 7 6 5 4 3 1
'''

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        i = n - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        if i == -1:
            nums[:] = nums[:: -1]
            return 
        
        # to find 5 pos
        j = n - 1
        while j >= 0 and nums[j] <= nums[i]:
            j -= 1
        # swap nums[i], nums[j]
        nums[i], nums[j] = nums[j], nums[i]
        # reverse from i+1 to n-1
        l, r = i + 1, n - 1
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l, r = l + 1, r - 1
